sensor/common/deploymentenhancer/component.go

```diff
@@ -0,0 +1,120 @@
+	return &DeploymentEnhancer{
+		responsesC:       make(chan *message.ExpiringMessage),
+		deploymentsQueue: make(chan *central.DeploymentEnhancementRequest, deploymentQueueSize),
+		storeProvider:    provider,
+	}
+}
+
+// ProcessMessage takes an incoming message and queues it for enhancement
+func (d *DeploymentEnhancer) ProcessMessage(msg *central.MsgToSensor) error {
+	toEnhance := msg.GetDeploymentEnhancementRequest()
+	if toEnhance == nil {
+		return nil
+	}
+	log.Debugf("Received message to process in DeploymentEnhancer: %+v", toEnhance)
+	d.deploymentsQueue <- toEnhance
+	return nil
+}
+
+// Start starts the component
+func (d *DeploymentEnhancer) Start() error {
+	go func() {
+		for {
+			deploymentMsg, more := <-d.deploymentsQueue
+			if !more {
+				return
+			}
+			requestID := deploymentMsg.GetMsg().GetId()
+			if requestID == "" {
+				log.Warnf("received deploymentEnhancement msg with empty request ID. Discarding request.")
+				continue
+			}
+			deployments := d.enhanceDeployments(deploymentMsg)
+			d.sendDeploymentsToCentral(requestID, deployments)
+		}
+	}()
+	return nil
+}
+
+func (d *DeploymentEnhancer) enhanceDeployments(deploymentMsg *central.DeploymentEnhancementRequest) []*storage.Deployment {
+	var ret []*storage.Deployment
+
+	deployments := deploymentMsg.GetMsg().GetDeployments()
+	if deployments == nil {
+		log.Warnf("received deploymentEnhancement message with no deployments")
+		return ret
+	}
+
+	log.Debugf("Received deploymentEnhancement msg with %d deployment(s)", len(deploymentMsg.GetMsg().GetDeployments()))
+	for _, deployment := range deployments {
+		d.enhanceDeployment(deployment)
+		ret = append(ret, deployment)
+	}
+
+	return ret
+}
+
+func (d *DeploymentEnhancer) sendDeploymentsToCentral(id string, deployments []*storage.Deployment) {
+	d.responsesC <- message.New(&central.MsgFromSensor{
+		Msg: &central.MsgFromSensor_DeploymentEnhancementResponse{
+			DeploymentEnhancementResponse: &central.DeploymentEnhancementResponse{
+				Msg: &central.DeploymentEnhancementMessage{
+					Id:          id,
+					Deployments: deployments,
+				},
+			},
+		},
+	})
+}
+
+func (d *DeploymentEnhancer) enhanceDeployment(deployment *storage.Deployment) {
+	d.storeProvider.Deployments().EnhanceDeploymentReadOnly(deployment, store.Dependencies{
+		PermissionLevel: d.storeProvider.RBAC().GetPermissionLevelForDeployment(deployment),
+		Exposures:       d.storeProvider.Services().GetExposureInfos(deployment.GetNamespace(), deployment.GetPodLabels()),
+	})
+}
+
+// Capabilities return the capabilities of this component
+func (d *DeploymentEnhancer) Capabilities() []centralsensor.SensorCapability {
+	// TODO(ROX-21197): Add Capability
+	return nil
+}
+
+// ResponsesC returns the response channel of this component
+func (d *DeploymentEnhancer) ResponsesC() <-chan *message.ExpiringMessage {
+	return d.responsesC
+}
+
+// Stop stops the component
+func (d *DeploymentEnhancer) Stop(_ error) {
+	defer close(d.deploymentsQueue)
```

Comment:
I think this is risky. Do we have a guarantee that nothing will write to that channel after the component is stopped? I guess this would require the components to be stopped in a particular order...

If this channel should live as long as Sensor, I would vote to keep it open forever (as we do in other components).

If this channel would remain open, then you will need a context (with cancel) to the object to finish the infinite loop in `Start()`.

So in `Start` add:

(pseudocode, untested)

```go
for {
    select {
        case <-d.ctx.Done():
            return
        case deploymentMsg, more := <-d.deploymentsQueue:
            // existing code
    }
}
```

and here in `Stop` add:

```go
func (d *DeploymentEnhancer) Stop(err error) {
	d.ctxCancel(err)
}
```

Reply:
>  Do we have a guarantee that nothing will write to that channel after the component is stopped?

No, because a message could be in the pipeline that was not yet passed to `ProcessMessage`.

> If this channel should live as long as Sensor, I would vote to keep it open forever (as we do in other components).

Agreed.

I think we should leave the channel open and stop the loop with a context like you suggested. `Stop` is never really called in production unless sensor is shutting down. Any potential risks of leaking the `Start` goroutine would be in tests only, but still should be fixed.

Reply:
I like the idea of using a context, this feels more safe. Done.